core/src/main/scala/kafka/server/KafkaApis.scala

```diff
@@ -417,38 +417,50 @@ class KafkaApis(val requestChannel: RequestChannel,
         }
       }
 
-      def produceResponseCallback(bandwidthThrottleTimeMs: Int) {
-        if (produceRequest.acks == 0) {
-          // no operation needed if producer request.required.acks = 0; however, if there is any error in handling
-          // the request, since no response is expected by the producer, the server will close socket server so that
-          // the producer client will know that some error has happened and will refresh its metadata
-          if (errorInResponse) {
-            val exceptionsSummary = mergedResponseStatus.map { case (topicPartition, status) =>
-              topicPartition -> status.error.exceptionName
-            }.mkString(", ")
-            info(
-              s"Closing connection due to error during produce request with correlation id ${request.header.correlationId} " +
-                s"from client id ${request.header.clientId} with ack=0\n" +
-                s"Topic and partition to exceptions: $exceptionsSummary"
-            )
-            closeConnection(request, new ProduceResponse(mergedResponseStatus.asJava).errorCounts)
+      // When this callback is triggered, the remote API call has completed
+      request.apiRemoteCompleteTimeNanos = time.nanoseconds
+
+      // Record both bandwidth and request quota-specific values and throttle by muting the channel if any of the quotas
+      // have been violated. If both quotas have been violated, use the max throttle time between the two quotas. Note
+      // that the request quota is not enforced if acks == 0.
+      val bandwidthThrottleTimeMs = quotas.produce.maybeRecord(request, numBytesAppended)
```

Comment:
Previously the way we throttle a request is pretty extensible: we first throttle based on byte rate, and if that passes, then we throttle based on produce rate. If in the future we have another throttle mechanism, e.g. based on CPU, we can easily stack this on top of byte-rate-based and request-rate-based throttling mechanism.

The way this patch implements throttling is kind of hard to extend (or appears to be difficult to read) due to the following reasons:

- The number of throttling mechanisms is hard-coded to be 2 in ThrottleChannel.tryUnmute()
- Instead of being able to distribute throttling mechanism in multiple places, now we have to determine the throttle time of all mechanism in one place (e.g. here) and use the one that has the largest throttle time. This will work for now. But it will be hard to extend and seems not-so-readable.
- We pass around Option[ThrottledChannel] and treats `None` separately from `Some()`. It will be cleaner to just pass one non-option object.

I am wondering if the following high-level solution would be better:

In class `KafkaChannel` we use an integer to keep track of the number of existing mechanisms that are throttling this channel. For example, if this request exceeds the byte-based quota, we increment this integer by one and enqueue an object into a delayed queue. After the corresponding throttle time is passed, we dequeue this object and decrement the value of this integer in the `KafkaChannel`. If the value is 0 after it is decremented, we unmute the channel.

This solution may be cleaner in the following sense:

- We no longer needs to pass Option[ThrottledChannel].
- We can implement byte-based and request-based throttling mechanism in the same manner as the existing Kafka implementation without having to put them together.
- The solution is more extensible since it does not assume there are exactly two throttling mechanisms.

What do you think?

Reply:
We discussed this offline and agreed to keep the current logic for using the max.

As for Option[ThrottledChannel], I used KafkaChannel to keep the ref count as suggested and got rid of the use of Option[ThrottledChannel].